File: Windows/FireflyDevice/FDFileLog.cpp
//
//  FDFileLog.cpp
//  FireflyDevice
//
//  Created by Denis Bohm on 12/23/13.
//  Copyright (c) 2013-2014 Firefly Design LLC / Denis Bohm. All rights reserved.
//

#include "FDFileLog.h"

#include <cstdio>
#include <iomanip>
#include <iostream>

namespace fireflydesign {

	FDFileLog::FDFileLog() {
		logLimit = 100000;
		logDirectory = std::tr2::sys::initial_path<std::tr2::sys::path>();
		logFileName = logDirectory / std::tr2::sys::path("log.txt");
		logFileNameOld = logDirectory / std::tr2::sys::path("log-1.txt");
	}

	std::string FDFileLog::getContent() {
		std::stringstream buffer;
		FDFileLog::getContent(buffer);
		return buffer.str();
	}

	void FDFileLog::getContent(std::stringstream& buffer) {
		std::lock_guard<std::mutex> lock(logMutex);

		close();

		appendFile(buffer, logFileNameOld);
		appendFile(buffer, logFileName);
	}

	void FDFileLog::appendFile(std::stringstream& buffer, std::string fileName) {
		std::ifstream stream(fileName);
		buffer << stream.rdbuf();
	}

	void FDFileLog::close() {
		logFile.close();
	}

	void FDFileLog::log(std::string message) {
		std::lock_guard<std::mutex> lock(logMutex);

		if (!logFile.is_open()) {
			logFile.open(logFileName, std::ofstream::out | std::ofstream::app);
			if (!logFile.is_open()) {
				std::tr2::sys::create_directories(logDirectory);
				logFile.open(logFileName, std::ofstream::out | std::ofstream::app);
			}
		}
		if (logFile.is_open()) {
			std::time_t time = std::time(NULL);
			struct tm tm;
			gmtime_s(&tm, &time);
			// "yyyy-MM-dd HH:mm:ss.SSS"
			logFile << std::put_time(&tm, "%Y-%m-%d %H:%M:%S") << " " << message << "\n";
			logFile.flush();
			unsigned long long length = logFile.tellp();
			if (length > logLimit) {
				close();
				std::tr2::sys::remove(logFileNameOld);
				std::tr2::sys::rename(logFileName, logFileNameOld);
			}
		}
	}

	std::string FDFileLog::lastPathComponent(std::string path) {
		std::size_t index = path.find_last_of("/\\");
		if (index != std::string::npos) {
			return path.substr(index + 1);
		}
		return path;
	}

	void FDFileLog::log(std::string file, unsigned line, std::string cls, std::string method, std::string message) {
		std::string fullMessage = lastPathComponent(file) + ":" + std::to_string(line) + " " + cls + "." + method + " " + message;

		std::cout << fullMessage << "\n";
		log(fullMessage);
	}

}

File: Android/Libraries/fireflydevice/src/main/java/com/fireflydesign/fireflydevice/FDFileLog.java
//
//  FDFileLog.cpp
//  FireflyDevice
//
//  Created by Denis Bohm on 12/23/13.
//  Copyright (c) 2013-2014 Firefly Design LLC / Denis Bohm. All rights reserved.
//

package com.fireflydesign.fireflydevice;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.PrintStream;

import java.util.Date;
import java.util.Scanner;

public class FDFileLog {

    public int logLimit;

    String logDirectory;
    String logFileName;
    String logFileNameOld;
    PrintStream logFile;

    public FDFileLog(String logDirectory) {
		logLimit = 100000;
		this.logDirectory = logDirectory;
		logFileName = logDirectory + "/log.txt";
		logFileNameOld = logDirectory + "/log-1.txt";
	}

	public String getContent() {
		StringBuffer buffer = new StringBuffer();
		getContent(buffer);
		return buffer.toString();
	}

	public synchronized void getContent(StringBuffer buffer) {
		close();

		appendFile(buffer, logFileNameOld);
		appendFile(buffer, logFileName);
	}

	static void appendFile(StringBuffer buffer, String fileName) {
        try {
            buffer.append(new Scanner(new File(fileName)).useDelimiter("\\Z").next());
        } catch (FileNotFoundException e) {
            System.out.println("can't open file: " + fileName);
        }
	}

	void close() {
		logFile.close();
        logFile = null;
	}

	synchronized void log(String message) {
		if (logFile == null) {
            try {
                logFile = new PrintStream(new FileOutputStream(logFileName, true));
			} catch (FileNotFoundException e) {
                System.out.println("can't open log file: " + logFileName);
            }
		}
		if (logFile != null) {
			String date = FDString.formatDateTime(new Date().getTime());
            logFile.println(date + " " + message);
			logFile.flush();
			long length = new File(logFileName).length();
			if (length > logLimit) {
				close();
				new File(logFileNameOld).delete();
				new File(logFileName).renameTo(new File(logFileNameOld));
			}
		}
	}

	String lastPathComponent(String path) {
		int index = path.lastIndexOf('/');
		if (index >= 0) {
			return path.substring(index + 1);
		}
		return path;
	}

	public void log(String file, int line, String cls, String method, String message) {
		String fullMessage = lastPathComponent(file) + ":" + Integer.toString(line) + " " + cls + "." + method + " " + message;

		System.out.println(fullMessage);
		log(fullMessage);
	}

}
